placement/placement.go

```diff
@@ -0,0 +1,667 @@
+func (sp storagePlacement) computeDesiredNumShards(
+	clusters map[string]*schema.Cluster,
+	totalShards int) map[string]int {
+
+	if len(clusters) == 0 {
+		return nil
+	}
+
+	// Compute the total weight of all clusters
+	var totalWeight uint32
+	for _, c := range clusters {
+		totalWeight += c.Weight
+	}
+
+	// compute desired # of shards based on relative weight of each cluster
+	var (
+		oldestCluster    *schema.Cluster
+		shardsRemaining  = totalShards
+		desiredNumShards = make(map[string]int, len(clusters))
+	)
+	for name, c := range clusters {
+		relativeWeight := (float64(c.Weight) / float64(totalWeight))
+		numShards := int(relativeWeight * float64(totalShards))
+		if numShards > shardsRemaining {
+			numShards = shardsRemaining
+		}
+
+		desiredNumShards[name] = numShards
+		shardsRemaining -= numShards
+
+		if oldestCluster == nil || oldestCluster.CreatedAt > c.CreatedAt {
+			oldestCluster = c
+		}
+	}
+
+	// if we haven't yet allocated all shards due to rounding, pick oldest
+	// cluster and add the remainder to it. we pick the oldest cluster to have a
+	// deterministic mapping and avoid moving shards around unnecessarily
+	if shardsRemaining > 0 {
+		desiredNumShards[oldestCluster.Name] += shardsRemaining
+	}
+	return desiredNumShards
+}
+
+func (sp storagePlacement) findDatabase(
+	p *schema.Placement,
+	changes *schema.PlacementChanges,
+	name string) (*schema.Database, *schema.DatabaseChanges) {
+	db := p.Databases[name]
+	if db == nil {
+		if add := changes.DatabaseAdds[name]; add != nil {
+			db = add.Database
+		}
+	}
+
+	if db == nil {
+		return nil, nil
+	}
+
+	dbChanges := changes.DatabaseChanges[name]
+	if dbChanges == nil {
+		dbChanges = &schema.DatabaseChanges{}
+		changes.DatabaseChanges[name] = dbChanges
+	}
+
+	return db, dbChanges
+}
+
+type placementUpdateFn func(*schema.Placement, *schema.PlacementChanges) error
+
+// wrapFn wraps a placement modification function
+func wrapFn(f func(*schema.Placement, *schema.PlacementChanges) error) func(proto.Message, proto.Message) error {
+	return func(protoConfig, protoChanges proto.Message) error {
+		var (
+			config  = protoConfig.(*schema.Placement)
+			changes = protoChanges.(*schema.PlacementChanges)
+		)
+
+		if config.Databases == nil {
+			config.Databases = make(map[string]*schema.Database)
+		}
+
+		if changes.DatabaseAdds == nil {
+			changes.DatabaseAdds = make(map[string]*schema.DatabaseAdd)
+		}
+
+		if changes.DatabaseChanges == nil {
+			changes.DatabaseChanges = make(map[string]*schema.DatabaseChanges)
+		}
+
+		return f(config, changes)
+	}
+}
+
+// storagePlacementOptions
+type storagePlacementOptions struct {
+	clock  clock.Clock
+	logger xlog.Logger
+}
+
+func (opts *storagePlacementOptions) GetClock() clock.Clock  { return opts.clock }
+func (opts *storagePlacementOptions) GetLogger() xlog.Logger { return opts.logger }
+
+func (opts *storagePlacementOptions) Clock(clock clock.Clock) StoragePlacementOptions {
+	opts.clock = clock
```

Comment:
This mutates options, do we want to follow using immutable options similar to M3DB? e.g.

```
type MyOptions interface { 
  GetThing() Thing
  Thing(value Thing)
}

type myOptions {
  thing Thing
}

func NewMyOptions() MyOptions {
  return &myOptions{thing: newDefaultThing()}
}

func (o *myOptions) Thing(value Thing) MyOptions {
  opts := *o
  opts.thing = value
  return &opts
}

func (o *myOptions) GetThing() Thing {
  return opts.thing
}
```